shared/entity/type.go

```diff
@@ -0,0 +1,332 @@
+	"github.com/canonical/lxd/shared/version"
+)
+
+// Type represents a resource type in LXD that is addressable via the API.
+type Type string
+
+// Reference uniquely identifies a LXD entity via its URL.
+type Reference string
+
+const (
+	// TypeContainer represents container resources.
+	TypeContainer Type = "container"
+
+	// TypeImage represents image resources.
+	TypeImage Type = "image"
+
+	// TypeProfile represents profile resources.
+	TypeProfile Type = "profile"
+
+	// TypeProject represents project resources.
+	TypeProject Type = "project"
+
+	// TypeCertificate represents certificate resources.
+	TypeCertificate Type = "certificate"
+
+	// TypeInstance represents instance resources.
+	TypeInstance Type = "instance"
+
+	// TypeInstanceBackup represents instance backup resources.
+	TypeInstanceBackup Type = "instance_backup"
+
+	// TypeInstanceSnapshot represents instance snapshot resources.
+	TypeInstanceSnapshot Type = "instance_snapshot"
+
+	// TypeNetwork represents network resources.
+	TypeNetwork Type = "network"
+
+	// TypeNetworkACL represents network acl resources.
+	TypeNetworkACL Type = "network_acl"
+
+	// TypeNode represents node resources.
+	TypeNode Type = "node"
+
+	// TypeOperation represents operation resources.
+	TypeOperation Type = "operation"
+
+	// TypeStoragePool represents storage pool resources.
+	TypeStoragePool Type = "storage_pool"
+
+	// TypeStorageVolume represents storage volume resources.
+	TypeStorageVolume Type = "storage_volume"
+
+	// TypeStorageVolumeBackup represents storage volume backup resources.
+	TypeStorageVolumeBackup Type = "storage_volume_backup"
+
+	// TypeStorageVolumeSnapshot represents storage volume snapshot resources.
+	TypeStorageVolumeSnapshot Type = "storage_volume_snapshot"
+
+	// TypeWarning represents warning resources.
+	TypeWarning Type = "warning"
+
+	// TypeClusterGroup represents cluster group resources.
+	TypeClusterGroup Type = "cluster_group"
+
+	// TypeStorageBucket represents storage bucket resources.
+	TypeStorageBucket Type = "storage_bucket"
+)
+
+const (
+	// pathPlaceholder is used to indicate that a path argument is expected in a URL.
+	pathPlaceholder = "{pathArgument}"
+)
+
+var entityTypes = []Type{
+	TypeContainer,
+	TypeImage,
+	TypeProfile,
+	TypeProject,
+	TypeCertificate,
+	TypeInstance,
+	TypeInstanceBackup,
+	TypeInstanceSnapshot,
+	TypeNetwork,
+	TypeNetworkACL,
+	TypeNode,
+	TypeOperation,
+	TypeStoragePool,
+	TypeStorageVolume,
+	TypeStorageVolumeBackup,
+	TypeStorageVolumeSnapshot,
+	TypeWarning,
+	TypeClusterGroup,
+	TypeStorageBucket,
+}
+
+// String implements fmt.Stringer for Type.
+func (t Type) String() string {
+	return string(t)
+}
+
+// Validate returns an error if the Type is not in the list of allowed types. If the allowEmpty argument is set to true
+// an empty string is allowed. This is to accommodate that warnings may not refer to a specific entity type.
+func (t Type) Validate(allowEmpty bool) error {
```

Comment:
Can we have the warnings subsystem decide whether or not to call Validate() based on whether its empty, rather than pushing that down into the generic package here?